concrete-core/src/math/decomposition/tests.rs

```diff
@@ -1,153 +1,134 @@
-use concrete_commons::CastInto;
+use crate::math::decomposition::{DecompositionBaseLog, DecompositionLevelCount, SignedDecomposer};
+use crate::math::random::{RandomGenerable, Uniform};
+use crate::math::tensor::Tensor;
+use crate::math::torus::UnsignedTorus;
+use crate::test_tools::{any_uint, any_usize, random_usize_between};
+use concrete_commons::{Numeric, SignedInteger, UnsignedInteger};
 use std::convert::TryInto;
-use std::fmt::Binary;
-
-use crate::crypto::UnsignedTorus;
-use crate::test_tools::{any_usize, any_utorus};
-
-use super::*;
-
-fn test_round_to_closest_multiple<T: UnsignedTorus>() {
-    let log_b = any_usize();
-    let level_max = any_usize();
-    let val = any_utorus::<T>();
-    let delta = any_utorus::<T>();
-    let bits = T::BITS;
-    let log_b = (log_b % ((bits / 4) - 1)) + 1;
-    let log_b: usize = log_b.try_into().unwrap();
-    let level_max = (level_max % 4) + 1;
-    let level_max: usize = level_max.try_into().unwrap();
-    let bit: usize = log_b * level_max;
-
-    let val = val << (bits - bit);
-    let delta = delta >> (bits - (bits - bit - 1));
+use std::fmt::Debug;
+
+// Returns a random decomposition valid for the size of the T type.
+fn random_decomp<T: UnsignedInteger>() -> SignedDecomposer<T> {
+    let mut base_log;
+    let mut level_count;
+    loop {
+        base_log = random_usize_between(2..T::BITS);
+        level_count = random_usize_between(2..T::BITS);
+        if base_log * level_count < T::BITS {
+            break;
+        }
+    }
+    SignedDecomposer::new(
+        DecompositionBaseLog(base_log),
+        DecompositionLevelCount(level_count),
+    )
+}
 
-    for _ in 0..1000 {
-        assert_eq!(
-            val,
-            val.wrapping_add(delta).round_to_closest_multiple(
-                DecompositionBaseLog(log_b),
-                DecompositionLevelCount(level_max)
-            )
-        );
-        assert_eq!(
-            val,
-            val.wrapping_sub(delta).round_to_closest_multiple(
-                DecompositionBaseLog(log_b),
-                DecompositionLevelCount(level_max)
-            )
-        );
+fn test_decompose_recompose<T: UnsignedInteger + Debug + RandomGenerable<Uniform>>()
+where
+    <T as UnsignedInteger>::Signed: Debug + SignedInteger,
+{
+    // Checks that the decomposing and recomposing a value brings the closest representable
+    for _ in 0..100_000 {
+        let decomposer = random_decomp::<T>();
+        let input = any_uint::<T>();
+        let closest = decomposer.closest_representable(input);
+        for term in decomposer.decompose(closest) {
```

Comment:
here input can be passed directly instead of closest. We could check that closest_rep(closest) = closest